src/ng-select/selection-model.ts

```diff
@@ -43,13 +45,19 @@ export class DefaultSelectionModel implements SelectionModel {
     unselect(item: NgOption, multiple: boolean) {
         this._selected = this._selected.filter(x => x !== item);
         item.selected = false;
+        item.indeterminate = false;
         if (multiple) {
             if (item.parent && item.parent.selected) {
                 const children = item.parent.children;
                 this._removeParent(item.parent);
                 this._removeChildren(item.parent);
                 this._selected.push(...children.filter(x => x !== item));
                 item.parent.selected = false;
+                item.parent.indeterminate = this._selected.length > 0 ? true : false;
+            } else if (item.parent) {
+                const childrenCount = item.parent.children.length;
+                const selectedCount = item.parent.children.filter(x => x.selected).length;
```

Comment:
Counting items like this is most inefficient because you are creating new array just to say how many selected items there are. You could use reduce or event simple loop;

```ts
const selectedCount = item.parent.children.reduce((c, n) => c + n.selected, 0);
```

Reply:
Hi I used the same line as 32 in the same file, maybe a change here is also possible?

Reply:
c is a number and n.selected boolean, this seems not to work. Should I keep the same code as for line 32 and you will refactor both line at the same time?